src/script/interpreter.cpp

```diff
@@ -61,6 +61,75 @@ static inline void popstack(std::vector<valtype>& stack)
     stack.pop_back();
 }
 
+static inline void push4_le(std::vector<valtype>& stack, uint32_t v)
+{
+    auto v_le = htole32(v);
+    valtype vch(reinterpret_cast<unsigned char*>(&v_le), reinterpret_cast<unsigned char*>(&v_le) + 4);
+    stack.push_back(vch);
+}
+
+static inline void push8_le(std::vector<valtype>& stack, uint64_t v)
+{
+    auto v_le = htole64(v);
+    valtype vch(reinterpret_cast<unsigned char*>(&v_le), reinterpret_cast<unsigned char*>(&v_le) + 8);
+    stack.push_back(vch);
+}
+
+static inline void pushasset(std::vector<valtype>& stack, const CConfidentialAsset& asset)
+{
+    assert(!asset.IsNull());
+    valtype vchinpAsset;
+    vchinpAsset.insert(vchinpAsset.begin(), asset.vchCommitment.begin() + 1, asset.vchCommitment.begin() + 33);
+    valtype vchAssetPref;
+    vchAssetPref.insert(vchAssetPref.begin(), asset.vchCommitment.begin(), asset.vchCommitment.begin() + 1);
+    stack.push_back(vchinpAsset);
+    stack.push_back(vchAssetPref);
+}
+
+static inline void pushvalue(std::vector<valtype>& stack, const CConfidentialValue& value)
+{
+    valtype vchinpValue;
+    if (value.IsNull()) {
+        vchinpValue.insert(vchinpValue.begin(), 8, 0);
+    } else if (value.IsExplicit()) {
+        // int64_t amt = value.GetAmount();
+        // Convert BE to LE by using reverse iterator
+        vchinpValue.insert(vchinpValue.begin(), value.vchCommitment.rbegin(), value.vchCommitment.rbegin() + 8);
+    } else { // (value.IsCommitment())
+        vchinpValue.insert(vchinpValue.begin(), value.vchCommitment.begin() + 1, value.vchCommitment.begin() + 33);
+    }
+    valtype vchValuePref;
+    if (!value.IsNull()) {
+        vchValuePref.insert(vchValuePref.begin(), value.vchCommitment.begin(), value.vchCommitment.begin() + 1);
+    } else {
+        // If value is null, explicitly push the explicit prefix 0x01
+        vchValuePref = valtype(1, 1);
+    }
+    stack.push_back(vchinpValue); // if value is null, 0(LE 8) is pushed
```

Comment:
Keeping in mind that I don't understand move semantics, my understanding from pieter is that we want std::move within these push_backs.